js/web/test/e2e/run.js

```diff
@@ -0,0 +1,106 @@
+const path = require('path');
+const fs = require('fs-extra');
+const globby = require('globby');
+const { spawn } = require('child_process');
+const startServer = require('./simple-http-server');
+
+// copy whole folder to out-side of <ORT_ROOT>/js/ because we need to test in a folder that no `package.json` file
+// exists in its parent folder.
+// here we use <ORT_ROOT>/build/js/e2e/ for the test
+
+const TEST_E2E_SRC_FOLDER = __dirname;
+const JS_ROOT_FOLDER = path.resolve(__dirname, '../../..');
+const TEST_E2E_RUN_FOLDER = path.resolve(JS_ROOT_FOLDER, '../build/js/e2e');
+const NPM_CACHE_FOLDER = path.resolve(TEST_E2E_RUN_FOLDER, '../npm_cache');
+const CHROME_USER_DATA_FOLDER = path.resolve(TEST_E2E_RUN_FOLDER, '../user_data');
+fs.emptyDirSync(TEST_E2E_RUN_FOLDER);
+fs.emptyDirSync(NPM_CACHE_FOLDER);
+fs.emptyDirSync(CHROME_USER_DATA_FOLDER);
+fs.copySync(TEST_E2E_SRC_FOLDER, TEST_E2E_RUN_FOLDER);
+
+// find packed package
+
+const ORT_COMMON_FOLDER = path.resolve(JS_ROOT_FOLDER, 'common');
+const ORT_COMMON_PACKED_FILEPATH_CANDIDATES = globby.sync('onnxruntime-common-*.tgz', { cwd: ORT_COMMON_FOLDER });
+if (ORT_COMMON_PACKED_FILEPATH_CANDIDATES.length !== 1) {
+  throw new Error('cannot find exactly single package for onnxruntime-common.');
+}
+const ORT_COMMON_PACKED_FILEPATH = path.resolve(ORT_COMMON_FOLDER, ORT_COMMON_PACKED_FILEPATH_CANDIDATES[0]);
+
+const ORT_WEB_FOLDER = path.resolve(JS_ROOT_FOLDER, 'web');
+const ORT_WEB_PACKED_FILEPATH_CANDIDATES = globby.sync('onnxruntime-web-*.tgz', { cwd: ORT_WEB_FOLDER });
+if (ORT_WEB_PACKED_FILEPATH_CANDIDATES.length !== 1) {
+  throw new Error('cannot find exactly single package for onnxruntime-web.');
+}
+const ORT_WEB_PACKED_FILEPATH = path.resolve(ORT_WEB_FOLDER, ORT_WEB_PACKED_FILEPATH_CANDIDATES[0]);
+
+// we start here:
+
+async function main() {
+  // install dev dependencies
+  await runInShell(`npm install"`);
+
+  // npm install with "--cache" to install packed packages with an empty cache folder
+  await runInShell(`npm install --cache "${NPM_CACHE_FOLDER}" "${ORT_COMMON_PACKED_FILEPATH}" "${ORT_WEB_PACKED_FILEPATH}"`);
+
+  // test case run in Node.js
+  await testAllNodejsCases();
+
+  // test cases with self-host (ort hosted in same origin)
+  await testAllBrowserCases({ hostInKarma: true });
+
+  // test cases without self-host (ort hosted in same origin)
+  startServer(path.resolve(TEST_E2E_RUN_FOLDER, 'node_modules', 'onnxruntime-web'));
+  await testAllBrowserCases({ hostInKarma: false });
+
+  // no error occurs, exit with code 0
+  process.exit(0);
+}
+
+async function testAllNodejsCases() {
+  await runInShell('node ./node_modules/mocha/bin/mocha ./node-test-main-no-threads.js');
+  await runInShell('node ./node_modules/mocha/bin/mocha ./node-test-main.js');
```

Comment:
this tests threads. does it pass the case without '--experimental-wasm-threads'?

Reply:
https://github.com/microsoft/onnxruntime/pull/7688/files#diff-43047bf857b508a7aedf9bf94689a3def19d8f1fd830832993e860d7b90cd42eR61-R64

we actually have all combinations (2x2=4) here.

`node-test-main-no-threads.js` means to force single thread
`node-test-main.js` means to use the default behavior. it does not necessarily use multi-thread -- the default behavior allows to fallback if threading is not supported.

Reply:
When node.js detects multiple logical cpus, it will check whether multi-threading is supported or not. Is a test for transferability of SABs working for node, too? Comment says it needed for Firefox. Except those two checks, do I miss other criteria to check multi-threading support?

Reply:
I don't think node has that check. I think the restriction is for security concerns in browser.